src/hpc/autoscale/cost/azurecost.py

```diff
@@ -0,0 +1,182 @@
+import csv
+import requests
+import hpc.autoscale.hpclogging as log
+from collections import namedtuple
+from requests_cache import CachedSession
+
+class azurecost:
+    def __init__(self, config: dict):
+
+        self.config = config
+        self.retail_url = "https://prices.azure.com/api/retail/prices?api-version=2021-10-01-preview&meterRegion='primary'"
+        self.clusters = config['cluster_name']
+        self.dimensions = namedtuple("dimensions", "cost,usage,region,meterid,meter,metercat,metersubcat,resourcegroup,tags,currency")
+        retail_name = f"{config['cache_root']}/retail"
+        self.retail_session = CachedSession(cache_name=retail_name,
+                                            backend='filesystem',
+                                            allowable_codes=(200,),
+                                            allowable_methods=('GET'),
+                                            expire_after=172800)
+
+        #_az_logger = logging.getLogger('azure.identity')
+        #_az_logger.setLevel(logging.ERROR)
+
+
+        ## If we have ACM data available use azcost format else use retail format.
+        ## for nodearray, combine usage data with either azcost or retail format.
+        self.DEFAULT_AZCOST_FORMAT="sku_name,region,spot,meter,meterid,metercat,metersubcat,resourcegroup,rate,currency"
+        self.RETAIL_FORMAT="sku_name,region,spot,meter,meterid,metercat,rate,currency"
+        self.NODEARRAY_USAGE_FORMAT="nodearray,core_hours,cost"
+
+        self.az_job_t = namedtuple('az_job_t', self.DEFAULT_AZCOST_FORMAT)
+        self.az_job_retail_t = namedtuple('az_job_retail_t', self.RETAIL_FORMAT)
+        self.az_array_t = namedtuple('az_array_t', (self.NODEARRAY_USAGE_FORMAT + ',' + self.DEFAULT_AZCOST_FORMAT))
+        self.az_array_retail_t = namedtuple('az_array_retail_t', (self.NODEARRAY_USAGE_FORMAT + ',' + self.RETAIL_FORMAT))
+
+    def do_meter_lookup(self, sku_name, spot, region):
```

Comment:
We try to use types everywhere in this codebase. 
i.e. something like

def do_meter_lookup(self, sku_name: str, spot: bool, region: str) -> Optional[float]:

where Optional is from the typing module

we use mypy to do the type checking.

Reply:
Fixed this one and a few other places, except where namedtuple is being returned. Leaving that as is for now.